packages/rpc-evm/src/adaptive-range-fetcher.ts

```diff
@@ -0,0 +1,137 @@
+import type { Cursor } from "@apibara/protocol";
+import type { FinalizedRangeResult } from "@apibara/protocol/rpc";
+import type { PublicClient } from "viem";
+import type { EvmRpcBlock } from "./block";
+import type { EvmRpcFilter } from "./filter";
+import { fetchLogsForRange } from "./log-fetcher";
+import { viemBlockHeaderToDna } from "./transform";
+
+const BLOCK_RANGE_ERROR_PATTERNS = [
+  "invalid block range params",
+  "block range",
+  "query returned more than",
+  "exceeds max",
+  "too many results",
+  "response size exceeded",
+  "log response size exceeded",
+  "result window",
+  "limit exceeded",
+] as const;
+
+export class AdaptiveRangeFetcher {
+  private currentBatchSize: bigint;
+  private minBatchSize: bigint;
+  private maxBatchSize: bigint;
+  private client: PublicClient;
+
+  constructor(
+    client: PublicClient,
+    initialBatchSize = 10n,
+    minBatchSize = 1n,
+    maxBatchSize = 1000n,
+  ) {
+    this.client = client;
+    this.currentBatchSize = initialBatchSize;
+    this.minBatchSize = minBatchSize;
+    this.maxBatchSize = maxBatchSize;
+  }
+
+  async fetchRange(
+    startBlock: bigint,
+    endBlock: bigint,
+    filters: EvmRpcFilter[],
+  ): Promise<FinalizedRangeResult<EvmRpcBlock>> {
+    const rangeEnd =
+      startBlock + this.currentBatchSize - 1n < endBlock
+        ? startBlock + this.currentBatchSize - 1n
+        : endBlock;
+
+    try {
+      const logsByBlock = await fetchLogsForRange(
+        this.client,
+        startBlock,
+        rangeEnd,
+        filters,
+      );
+
+      const newSize = this.currentBatchSize * 2n;
+      this.currentBatchSize =
+        newSize > this.maxBatchSize ? this.maxBatchSize : newSize;
+
+      const blockPromises = [];
+      for (let bn = startBlock; bn <= rangeEnd; bn++) {
```

Comment:
Notice that we should not fetch the blocks for blocks that don't include any log (unless the header filter is "always"). You can elegantly create a `blockRange({ filters, logs })` generator to generate the block range based on both the provided filter (looking at the header inclusion flag) and the returned logs.